app/src/org/commcare/provider/InstanceProvider.java

```diff
@@ -542,10 +542,14 @@ private void linkToSessionFormRecord(Uri instanceUri) {
             // before moving on. We'll catch any errors here and just eat them
             // (since the task will also try the process and fail if it does.
             if (FormRecord.STATUS_COMPLETE.equals(current.getStatus())) {
+                net.sqlcipher.database.SQLiteDatabase db =
+                        CommCareApplication.instance().getUserDbHandle();
+                db.beginTransaction();
                 try {
                     new FormRecordProcessor(getContext()).process(current);
```

Comment:
I wonder if it would make more sense to move the transactionality here inside of the process event rather than here? It doesn't seem like the side effects of what happens in this method will change, and generally speaking it is ~always the right decision for an individual form to process transactionally, so there's no reason why we wouldn't want to change how process() works to protect the ability to do a non-transactional form process

Reply:
@ctsims I'm not sure how that would work. In order to ensure that `endTransaction()` gets called regardless of any exceptions encountered, that call _has_ to be made outside of the `process()` method itself (as is happening in the `finally` block here). While I theoretically could move the `beginTransaction()` and `setTransactionSuccessful` calls to _within_ `process()`, it seems worse to me to have different portions of the transaction-related code in different places; I would much rather it be getting clearly handled in one place. Let me know if I'm missing some other way of acheiving this.